crates/tx5/src/backend.rs

```diff
@@ -0,0 +1,134 @@
+//! Backend modules usable by tx5.
+
+use std::io::Result;
+use std::sync::Arc;
+
+use futures::future::BoxFuture;
+
+use crate::{Config, PubKey};
+use tx5_core::deps::serde_json;
+
+#[cfg(feature = "backend-go-pion")]
+mod go_pion;
+
+/// Backend modules usable by tx5.
+#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Hash)]
+pub enum BackendModule {
+    #[cfg(feature = "backend-go-pion")]
+    /// The Go Pion-based backend.
+    GoPion,
+
+    #[cfg(feature = "backend-webrtc-rs")]
+    /// The Webrtc-RS-based backend.
+    WebrtcRs,
+
+    /// The mock backend.
+    Mock,
+}
+
+impl Default for BackendModule {
+    #[allow(unreachable_code)]
+    fn default() -> Self {
+        #[cfg(feature = "backend-go-pion")]
+        return Self::GoPion;
+        #[cfg(feature = "backend-webrtc-rs")]
+        return Self::WebrtcRs;
+        Self::Mock
+    }
+}
+
+impl BackendModule {
+    /// Get a default version of the module-specific config.
+    pub fn default_config(&self) -> serde_json::Value {
+        match self {
+            #[cfg(feature = "backend-go-pion")]
+            Self::GoPion => go_pion::default_config(),
+            #[cfg(feature = "backend-webrtc-rs")]
+            Self::WebrtcRs => todo!(),
+            Self::Mock => serde_json::json!({}),
+        }
+    }
+
+    /// Connect a new backend module endpoint.
+    pub async fn connect(
+        &self,
+        url: &str,
+        listener: bool,
+        config: &Arc<Config>,
+    ) -> Result<(DynBackEp, DynBackEpRecv)> {
+        match self {
+            #[cfg(feature = "backend-go-pion")]
+            Self::GoPion => go_pion::connect(config, url, listener).await,
+            #[cfg(feature = "backend-webrtc-rs")]
+            Self::WebrtcRs => todo!(),
+            Self::Mock => todo!(),
+        }
+    }
+}
+
+/// Backend connection.
+pub trait BackCon: 'static + Send + Sync {
+    /// Send data over this backend connection.
+    fn send(&self, data: Vec<u8>) -> BoxFuture<'_, Result<()>>;
+
+    /// Get the pub_key identifying this connection.
+    fn pub_key(&self) -> &PubKey;
+
+    /// Returns `true` if we successfully connected over webrtc.
+    // TODO - this isn't good encapsulation
+    fn is_using_webrtc(&self) -> bool;
```

Comment:
What would the right interface look like here? Something like an `inner` that let's you get at the real connection and you have to know which ones it makes sense to ask about the use of webrtc as the caller?

Reply:
It's the latter. "Connections" can communicate directly through sbd, then they attempt to upgrade to a webrtc connection. So `is_using_webrtc()` returns true once/if that upgrade is successful.

Reply:
> What would the right interface look like here?

Yeah, that's still TBD. Does it even make sense to have this call be separate? Maybe it should just be part of the stats call, although that would just defer the problem, because there's an encapsulation question about that too : )

Reply:
Maybe that's a different topic, but why does tx5 need to know whether it's connected indirectly through sbd or directly through webrtc?

Reply:
As far as I remember, it's informational or for unit tests. But I still submit this is something that doesn't have to be sorted out to merge this PR, since I'm not changing any functionality here.

Reply:
I asked out of curiosity, it wasn't meant as a subtle push to change it. I prefer being explicit about pushes ;-) Fine by me to keep as is.